### ISSUE TITLE
Broadcast blocks don't work as expected (help wanted!)

### ISSUE THREAD

**Issue by PullJosh** [PullJosh/scratch-js#10]
Right now, the `broadcast` and `broadcastAndWait` blocks don't work correctly when they call each other. The goal is to perfectly replicate Scratch's behavior.

I've been struggling to make it work myself. If someone could help debug this one, I'd be thrilled.

(Fork from the [broadcasts](https://github.com/PullJosh/scratch-js/tree/broadcasts) branch and then submit a PR.)

**Comment by tjvr** [comment]
Have you found the "spec" of how they should behave? I knew this once and could probably dig it out...

**Comment by PullJosh** [comment]
<sup>(Sorry for the slow response; life's busy.)</sup>
From what I can discern, this is how the event loop works in Scratch:
* There is a list in memory of currently running scripts, and the position of execution in the code. (In scratch-js we use [generator functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*) for this because it allows us to yield and then resume execution while retaining the old context.)
* Each frame, the interpreter iterates over all the currently executing scripts (in a deterministic order -- not terribly concerned about exactly what order that is, at least for now). Every script runs for a while, until it hits a stopping point (done with the `yield` keyword in scratch-js.)
  * There are a few different ways that scripts can yield:
    * The end of a loop is reached (repeat, forever, repeat until). Yields one time, then continues.
    * A wait block is reached (wait, wait until, ask and wait, say/think and wait, play sound and wait, broadcast and wait, plus the glide block -- not technically a wait, but similar). These yield over and over, which means that they don't do anything when the scripts are iterated by the interpreter, until the wait condition is finished. In scratch-js, we use `yield*` for this.
      * The broadcast and wait block only yields if there is a non-empty receiving script which fires as a result of the broadcast being sent. I'm not worried about performing the non-empty check in scratch-js. If the broadcast triggers any events to be fired, it can yield.
* A script can only ever be running once at a time (per sprite -- clones can all run the same script at the same time, each for their own instance).
  * If a script is called again (its event<sup>1</sup> fires) while it is running, things get tricky.
    * If the event is a `when green flag clicked`, `when broadcast received`, `when this sprite clicked`, or `when backdrop switches`, it cancels the previous execution and restarts from the top.
    * If the event is a `when key pressed` or `when [sensor] > ()` the event is ignored completely.
    * Clicking on a script in the editor while it is running will cause it to stop execution without restarting. Not a concern for scratch-js.
   * Nothing is different about a script triggering itself (broadcast in a receive, restart timer in a timer event, switching backdrop in a backdrop switches). A broadcast will cancel itself (even if a `broadcast and wait` block is used), a backdrop switch will cancel itself, and a `when [sensor] > ()` will continue without being triggered again.

1: We call events "triggers" in scratch-js because [`Event`](https://developer.mozilla.org/en-US/docs/Web/API/Event) was already taken

---

As always, writing this out has helped me to understand it myself. Still, I'm feeling a little overwhelmed by the bugs. Help would still be appreciated, if you have the time. :wink: If it helps, [this is a copy](https://scratch.mit.edu/projects/260955918/) of the example project so you can see how it should actually work.

**Comment by tjvr** [comment]
Yeah, that sounds about right, from what I can remember!

---

Phosphorus cancels the currently-running script if it broadcasts itself:
```js
      } else if (block[0] === 'broadcast:') { /* Control */

        source += 'var threads = broadcast(' + val(block[1]) + ');\n';
        source += 'if (threads.indexOf(BASE) !== -1) {return;}\n';
```

It does the same thing for broadcast and wait (ignore the extra `R.` here, that's just used to store context that can survive across a yield).
```js
        source += 'R.threads = broadcast(' + val(block[1]) + ');\n';
        source += 'if (R.threads.indexOf(BASE) !== -1) {return;}\n';
```

It then waits until all of the  "when I receive" blocks have finished running.
```js
        var id = label();
        source += 'if (running(R.threads)) {\n';
        forceQueue(id);
        source += '}\n';
```

**Comment by PullJosh** [comment]
Fixed in #13
